File: ios/BFExplorer/Plaoc/IPCCore/IPC/IpcRequest.swift
//
//  IpcRequest.swift
//  IPC
//
//  Created by ui03 on 2023/2/13.
//

import UIKit
import Vapor

class IpcRequest {

    var type: IPC_MESSAGE_TYPE = .REQUEST
    var parsed_url: URL?
    var req_id: Int = 0
    var method: String = ""
    var headers: IpcHeaders?
    var urlString: String = ""
    var body: IpcBody?
    var ipc: Ipc?
    var url: URL?
    
    required init() {
        
    }
    
    init(req_id: Int, method: String, urlString: String, headers: IpcHeaders, body: IpcBody?, ipc: Ipc?) {
        
        self.ipc = ipc
        self.urlString = urlString
        self.req_id = req_id
        self.method = method
        self.headers = headers
        self.body = body
        url = URL(string: urlString)
        
        if let sender = body as? IpcBodySender, ipc != nil {
            IpcBodySender().IPCsender(ipc: ipc!, ipcBody: sender)
        }
    }
    
    static func fromText(text: String,req_id: Int,method: String = "GET",urlString: String,headers: IpcHeaders, ipc: Ipc) -> IpcRequest {
        return IpcRequest(req_id: req_id, method: method, urlString: urlString, headers: headers, body: IpcBodySender.from(raw: text, ipc: ipc), ipc: ipc)
    }
    
    static func fromBinary(binary: [UInt8],req_id: Int,method: String,urlString: String,headers:IpcHeaders, ipc: Ipc) -> IpcRequest {
        
        headers.set(key: "Content-Type", value: "application/octet-stream")
        headers.set(key: "Content-Length", value: "\(binary.count)")
        
        return IpcRequest(req_id: req_id, method: method, urlString: urlString, headers: headers, body: IpcBodySender.from(raw: binary, ipc: ipc), ipc: ipc)
    }
    
    static func fromStream(stream: InputStream,req_id: Int,method: String,urlString: String,headers:IpcHeaders, ipc: Ipc, size: Int64?) -> IpcRequest {
        
        headers.set(key: "Content-Type", value: "application/octet-stream")
        if size != nil {
            headers.set(key: "Content-Length", value: "\(size!)")
        }
        return IpcRequest(req_id: req_id, method: method, urlString: urlString, headers: headers, body: IpcBodySender.from(raw: stream, ipc: ipc), ipc: ipc)
    }
    
    static func fromRequest(req_id: Int,request: Request, ipc: Ipc) -> IpcRequest {
        
        let header = IpcHeaders(content: request.headers.description)
        
        if request.method == .GET || request.method == .HEAD {
            return IpcRequest(req_id: req_id, method: request.method.rawValue, urlString: request.url.string, headers: header, body: IpcBodySender.from(raw: "", ipc: ipc), ipc: ipc)
        }
        
        if request.body.data != nil {
            return self.fromBinary(binary: [UInt8](Data(buffer: request.body.data!)), req_id: req_id, method: request.method.rawValue, urlString: request.url.string, headers: header, ipc: ipc)
        } else if request.method == .POST || request.method == .PUT || request.method == .PATCH {
            var ipc_req_body_stream: Data = Data()
            var sequential = request.eventLoop.makeSucceededFuture(())
            
            request.body.drain {
                switch $0 {
                case .buffer(var buffer):
                    if buffer.readableBytes > 0 {
                        ipc_req_body_stream.append(buffer.readData(length: buffer.readableBytes)!)
                    }
                    
                    return sequential
                case .error(_):
                    return sequential
                case .end:
                    return sequential
                }
            }
            
            return self.fromStream(stream: InputStream(data: ipc_req_body_stream), req_id: req_id, method: request.method.rawValue, urlString: request.url.string, headers: header, ipc: ipc, size: Int64(ipc_req_body_stream.count))
        }else {
            return self.fromText(text: request.body.string ?? "", req_id: req_id, method: request.method.rawValue, urlString: request.url.string, headers: header, ipc: ipc)
        }
    }
    
    func toRequest() -> Request {
        
        if method == "GET" || method == "HEAD" {
            return Request.new(method: HTTPMethod(rawValue: method), url: self.urlString)
        }
        
        var buffer: ByteBuffer
        
        if let content = body?.raw as? String {
            buffer = .init(string: content)
        } else if let bytes = body?.raw as? [UInt8] {
            buffer = .init(data: Data(bytes))
        } else if let stream = body?.raw as? InputStream {
            buffer = .init(data: Data(stream.readByteArray()))
        } else {
            fatalError("invalid body to request: \(body)")
        }
        return Request.new(method: HTTPMethod(rawValue: method), url: self.urlString, collectedBody: buffer)
        
    }
    
    func toString() -> String {
        return "#IpcRequest/\(method)/\(urlString)"
    }
    
    lazy var ipcReqMessage: IpcReqMessage = {
        return IpcReqMessage(req_id: req_id, method: method, urlString: urlString, headers: headers?.headerDict ?? [:], metaBody: body?.metaBody)
    }()
}

extension IpcRequest: IpcMessage { }

File: next/ts/electron/esm/core/ipc/IpcRequest.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _IpcRequest_parsed_url;
import { once } from "lodash";
import { binaryToU8a, isBinary } from "../../helper/binaryHelper.js";
import { parseUrl } from "../../helper/urlHelper.js";
import { IpcMessage, IPC_MESSAGE_TYPE, IPC_METHOD, toIpcMethod, } from "./const.js";
import { IpcBodySender } from "./IpcBodySender.js";
import { IpcHeaders } from "./IpcHeaders.js";
export class IpcRequest extends IpcMessage {
    constructor(req_id, url, method, headers, body, ipc) {
        super(IPC_MESSAGE_TYPE.REQUEST);
        Object.defineProperty(this, "req_id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: req_id
        });
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: url
        });
        Object.defineProperty(this, "method", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: method
        });
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: headers
        });
        Object.defineProperty(this, "body", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: body
        });
        Object.defineProperty(this, "ipc", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ipc
        });
        _IpcRequest_parsed_url.set(this, void 0);
        Object.defineProperty(this, "ipcReqMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: once(() => new IpcReqMessage(this.req_id, this.method, this.url, this.headers.toJSON(), this.body.metaBody))
        });
        if (body instanceof IpcBodySender) {
            IpcBodySender.$usableByIpc(ipc, body);
        }
    }
    get parsed_url() {
        return (__classPrivateFieldSet(this, _IpcRequest_parsed_url, __classPrivateFieldGet(this, _IpcRequest_parsed_url, "f") ?? parseUrl(this.url), "f"));
    }
    static fromText(req_id, url, method = IPC_METHOD.GET, headers = new IpcHeaders(), text, ipc) {
        // 这里 content-length 默认不写，因为这是要算二进制的长度，我们这里只有在字符串的长度，不是一个东西
        return new IpcRequest(req_id, url, method, headers, IpcBodySender.from(text, ipc), ipc);
    }
    static fromBinary(req_id, url, method = IPC_METHOD.GET, headers = new IpcHeaders(), binary, ipc) {
        headers.init("Content-Type", "application/octet-stream");
        headers.init("Content-Length", binary.byteLength + "");
        return new IpcRequest(req_id, url, method, headers, IpcBodySender.from(binaryToU8a(binary), ipc), ipc);
    }
    static fromStream(req_id, url, method = IPC_METHOD.GET, headers = new IpcHeaders(), stream, ipc) {
        headers.init("Content-Type", "application/octet-stream");
        return new IpcRequest(req_id, url, method, headers, IpcBodySender.from(stream, ipc), ipc);
    }
    static fromRequest(req_id, ipc, url, init = {}) {
        const method = toIpcMethod(init.method);
        const headers = init.headers instanceof IpcHeaders
            ? init.headers
            : new IpcHeaders(init.headers);
        let ipcBody;
        if (isBinary(init.body)) {
            ipcBody = IpcBodySender.from(init.body, ipc);
        }
        else if (init.body instanceof ReadableStream) {
            ipcBody = IpcBodySender.from(init.body, ipc);
        }
        else {
            ipcBody = IpcBodySender.from(init.body ?? "", ipc);
        }
        return new IpcRequest(req_id, url, method, headers, ipcBody, ipc);
    }
    toRequest() {
        const { method } = this;
        let body;
        if ((method === IPC_METHOD.GET || method === IPC_METHOD.HEAD) === false) {
            body = this.body.raw;
        }
        const init = {
            method,
            headers: this.headers,
            body,
        };
        if (body) {
            Reflect.set(init, "duplex", "half");
        }
        return new Request(this.url, init);
    }
    toJSON() {
        const { method } = this;
        let body;
        if ((method === IPC_METHOD.GET || method === IPC_METHOD.HEAD) === false) {
            body = this.body.raw;
            return new IpcReqMessage(this.req_id, this.method, this.url, this.headers.toJSON(), this.body.metaBody);
        }
        return this.ipcReqMessage();
    }
}
_IpcRequest_parsed_url = new WeakMap();
export class IpcReqMessage extends IpcMessage {
    constructor(req_id, method, url, headers, metaBody) {
        super(IPC_MESSAGE_TYPE.REQUEST);
        Object.defineProperty(this, "req_id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: req_id
        });
        Object.defineProperty(this, "method", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: method
        });
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: url
        });
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: headers
        });
        Object.defineProperty(this, "metaBody", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: metaBody
        });
    }
}
